src/NamespaceListing.elm

```diff
@@ -0,0 +1,111 @@
+module NamespaceListing exposing
+    ( DefinitionListing(..)
+    , NamespaceListing(..)
+    , NamespaceListingContent
+    , decode
+    , map
+    )
+
+import FullyQualifiedName exposing (FQN, fqn)
+import Hash exposing (Hash(..))
+import Json.Decode as Json exposing (andThen, field)
+import List.Nonempty
+import RemoteData exposing (RemoteData(..), WebData)
+
+
+type DefinitionListing
+    = Type Hash FQN
+    | Term Hash FQN
+    | Patch String
+
+
+type alias NamespaceListingContent =
+    { definitions : List DefinitionListing, namespaces : List NamespaceListing }
+
+
+type NamespaceListing
+    = NamespaceListing Hash FQN (WebData NamespaceListingContent)
+
+
+map : (NamespaceListing -> NamespaceListing) -> NamespaceListing -> NamespaceListing
+map f ((NamespaceListing hash fqn content) as namespace) =
+    let
+        mapContent c =
+            { c | namespaces = List.map (map f) c.namespaces }
+    in
+    f (NamespaceListing hash fqn (RemoteData.map mapContent content))
+
+
+
+-- JSON Decode
+
+
+decode : Json.Decoder NamespaceListing
+decode =
+    Json.map3
+        NamespaceListing
+        (field "namespaceListingHash" Json.string |> andThen (Hash >> Json.succeed))
+        (field "namespaceListingName" Json.string |> andThen (fqn >> Json.succeed))
+        -- The main namespace being decoded has children, so we use Success for
+        -- the RemoteData. There children of the children however are not yet
+        -- fetched
+        (field "namespaceListingChildren" decodeContent |> andThen (Success >> Json.succeed))
+
+
+
+-- JSON Decode Helpers
+
+
+{-| Decoding specific intermediate type |
+-}
+type DecodedNamespaceChild
+    = SubNamespace NamespaceListing
+    | SubDefinition DefinitionListing
+
+
+decodeSubNamespace : Json.Decoder DecodedNamespaceChild
+decodeSubNamespace =
+    Json.map3 NamespaceListing
+        -- TODO namespaceName should be namespaceHash
+        (field "namespaceName" Json.string |> andThen (Hash >> Json.succeed))
```

Comment:
@runarorama would it be possible to had both an fqn and a hash to any NamespaceListingChild?